storage: TestPebbleMapClose failed

[(storage).TestPebbleMapClose failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2109474&tab=buildLog) on [master@dafb69aa4af50d965698738323680c9175d59511](https://github.com/cockroachdb/cockroach/commits/dafb69aa4af50d965698738323680c9175d59511):

```

6:  - 

6:  - 

6:  - 

6:  - 

6:  - 

6:  - 

6:  - 

6:  - 

6:  - 

6:  - 

6:  - 

6:  - 

6:  - 

6:  - 

6:  - 

6:  - 

6:  - 
    TestPebbleMapClose: disk_map_test.go:510: condition failed to evaluate within 45s: 
        6:  - 
        
        goroutine 262 [running]:
        runtime/debug.Stack(0xc000679d50, 0x2da9c40, 0xc000119be0)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xab
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x2df2040, 0xc00061c000, 0xc000679d50)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:36 +0x87
        github.com/cockroachdb/cockroach/pkg/storage.TestPebbleMapClose(0xc00061c000)
        	/go/src/github.com/cockroachdb/cockroach/pkg/storage/disk_map_test.go:510 +0xb7c
        testing.tRunner(0xc00061c000, 0x2b25a08)
        	/usr/local/go/src/testing/testing.go:991 +0x1ec
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1042 +0x661
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestPebbleMapClose884009224
--- FAIL: TestPebbleMapClose (45.14s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestPebbleMapClose PKG=./pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestPebbleMapClose.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/storage/disk_map_test.go
+++ b/pkg/storage/disk_map_test.go
@@ -440,37 +440,6 @@ func TestPebbleMapClose(t *testing.T) {
 	}
 	defer e.Close()
 
-	decodeKey := func(v []byte) []byte {
-		var err error
-		v, _, err = encoding.DecodeUvarintAscending(v)
-		if err != nil {
-			t.Fatal(err)
-		}
-		return v
-	}
-
-	getSSTables := func() string {
-		var buf bytes.Buffer
-		fmt.Fprintf(&buf, "\n")
-		for l, ssts := range e.db.SSTables() {
-			for _, sst := range ssts {
-				fmt.Fprintf(&buf, "%d: %s - %s\n",
-					l, decodeKey(sst.Smallest.UserKey), decodeKey(sst.Largest.UserKey))
-			}
-		}
-		return buf.String()
-	}
-
-	verifySSTables := func(expected string) {
-		actual := getSSTables()
-		if expected != actual {
-			t.Fatalf("expected%sgot%s", expected, actual)
-		}
-		if testing.Verbose() {
-			fmt.Printf("%s", actual)
-		}
-	}
-
 	diskMap := newPebbleMap(e.db, false /* allowDuplicates */)
 
 	// Put a small amount of data into the disk map.
@@ -498,22 +467,58 @@ func TestPebbleMapClose(t *testing.T) {
 	}
 
 	// Verify we have a single sstable.
-	verifySSTables(`
-6: a - z
-`)
+	var buf bytes.Buffer
+	fileNums := map[pebble.FileNum]bool{}
+	for l, ssts := range e.db.SSTables() {
+		for _, sst := range ssts {
+			sm := bytes.TrimPrefix(sst.Smallest.UserKey, diskMap.prefix)
+			la := bytes.TrimPrefix(sst.Largest.UserKey, diskMap.prefix)
+			fmt.Fprintf(&buf, "%d: %s - %s\n", l, sm, la)
+			fileNums[sst.FileNum] = true
+		}
+	}
+	const expected = "6: a - z\n"
+	actual := buf.String()
+	if expected != actual {
+		t.Fatalf("expected\n%sgot\n%s", expected, actual)
+	}
+	if testing.Verbose() {
+		fmt.Printf("%s", actual)
+	}
 
 	// Close the disk map. This should both delete the data, and initiate
 	// compactions for the deleted data.
 	diskMap.Close(ctx)
 
-	// Wait for the data stored in the engine to disappear.
+	// Wait for the previously-observed sstables to be removed. We can't
+	// assert that the LSM is completely empty because the range tombstone's
+	// sstable will not necessarily be compacted.
+	var lsmBuf bytes.Buffer
+	var stillExistBuf bytes.Buffer
 	testutils.SucceedsSoon(t, func() error {
-		actual := getSSTables()
-		if testing.Verbose() {
-			fmt.Printf("%s", actual)
+		lsmBuf.Reset()
+		stillExistBuf.Reset()
+		for l, ssts := range e.db.SSTables() {
+			if len(ssts) == 0 {
+				continue
+			}
+
+			fmt.Fprintf(&lsmBuf, "L%d:\n", l)
+			for _, sst := range ssts {
+				if fileNums[sst.FileNum] {
+					fmt.Fprintf(&stillExistBuf, "%s\n", sst.FileNum)
+				}
+				fmt.Fprintf(&lsmBuf, "  %s: %d bytes, %x (%s) - %x (%s)\n", sst.FileNum, sst.Size,
+					sst.Smallest.UserKey, bytes.TrimPrefix(sst.Smallest.UserKey, diskMap.prefix),
+					sst.Largest.UserKey, bytes.TrimPrefix(sst.Largest.UserKey, diskMap.prefix))
+			}
 		}
-		if actual != "\n" {
-			return fmt.Errorf("%s", actual)
+
+		if testing.Verbose() {
+			fmt.Println(buf.String())
+		}
+		if stillExist := stillExistBuf.String(); stillExist != "" {
+			return fmt.Errorf("%s", stillExist)
 		}
 		return nil
 	})

